Performance optimization for HillClimbSearch [fixes #939]

ankurankan opened the pull request:
Since hill climb search algorithm makes a lot of calls to BaseEstimator.state_counts
with the same arguments and computing them all makes them quite slow. So, this commit
adds a lru_cache on it, so that we don't have to compute the values everytime. The
default value of cache size is 2048 right now, but in future we would like to have
a dynamic size allocation based on system memory. Cache of this size is takes around
1.5 GB of memory and should have 100% hits (except for first computations) on learning
models of 12-15 variables.

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.

### Fixes #939 . (if there is no issue for this, please create one)

#### Changes
1. Adds caching to `BaseEstimator.state_counts`
2. Fixes tests to pass tuples instead of lists as caching would require them to be hashable.
3. Adds `backports.lru_cache` in the extern module: https://pypi.python.org/pypi/backports.functools_lru_cache/1.0.1

💔Thank you!

ankurankan committed:
Performance optimization for HillClimbSearch [fixes #939]:

Since hill climb search algorithm makes a lot of calls to BaseEstimator.state_counts
with the same arguments and computing them all makes them quite slow. So, this commit
adds a lru_cache on it, so that we don't have to compute the values everytime. The
default value of cache size is 2048 right now, but in future we would like to have
a dynamic size allocation based on system memory. Cache of this size is takes around
1.5 GB of memory and should have 100% hits (except for first computations) on learning
models of 12-15 variables.

This is a combination of 2 commits.
This is the 1st commit message:

adds lru cache

This is the commit message #2:

fix tests

ankurankan committed:
lru_cache support for python versions < 3.3:

Changes:
1. Adds backports.functools_lru_cache as an external module.
2. While importing checks whether the default implementation is present (py >= 3.3) otherwise fallsback to the external module.

khalibartan commented on pgmpy/estimators/base.py:
```diff
@@ -1,5 +1,9 @@
 #!/usr/bin/env python
 from warnings import warn
+try:
+    from functools import lru_cache
```
@ankurankan functools lru_cache is as fast the attached lru_cache. Actually I think implementation is same if you checkout the lru_cache implementation in functools.

ankurankan replied:
@khalibartan Yeah, but that might change in newer versions of python. Plus things might also break in `backports.lru_cache` for newer versions of python. And I think it's better to always use the default implementation if possible because python community will be fixing any new issues/bugs whereas we can't have that guarantee for third party modules.

ankurankan committed:
Few changes in caching and adds extra parameters to estimate:

1. Adds parameters epsilon and max_iter to estimate method.
2. Also cache local_score, so that it doesn't call state_names everytime.

minor changes to fix tests

ankurankan committed:
try parallelizing

khalibartan commented:
@ankurankan Status ??

ankurankan committed:
Adds joblib to requirements

ankurankan commented:
@khalibartan This doesn't exactly solve #939 but is a huge performance improvement. I will clean up the PR a bit and then we can merge this I think. And I will come back to #939 once the continuous variables are implemented.

ankurankan committed:
adds tqdm to requirements

senthilki commented:
Am getting below error after applying all six commits for performance improvement.

===============================================
KeyError                                  Traceback (most recent call last)
<ipython-input-5-9368fa650cd9> in <module>()
     47 #from pgmpy.estimators import HillClimbSearch
     48 hc = HillClimbSearch(sluggish, scoring_method=BicScore(sluggish))
---> 49 best_model = hc.estimate()
     50 print(best_model.edges())
     51 

C:\Program Files\Anaconda3\lib\site-packages\pgmpy\estimators\HillClimbSearch.py in estimate(self, start, tabu_length, max_indegree, epsilon, max_iter)
    217 
    218             r = self._legal_operations(p, current_model, tabu_list, max_indegree)
--> 219             best_operation = (op_dict[r[0]], (r[2], r[3]))
    220             best_score_delta = float(r[1])
    221 

KeyError: '2'
================================

Also can you please help me in understanding below queries ? 

1. Does Pgmpy supports continuous variable ? 
2. Pgmpy taking more than 7 hrs for drawing the model , whereas one of my friend using "R" and "Hillclimbsearch" is taking less than 1 minute to draw bn with more accuracy , I felt it is related to my question 1.

Please help me to resolve the above issue.

ankurankan commented:
@senthilki Thanks for reporting the issue. The PR is not ready yet and that is a know error. 

> Does Pgmpy supports continuous variable ?

No we don't have complete support for continuous variables yet.

> Pgmpy taking more than 7 hrs for drawing the model , whereas one of my friend using "R" and "Hillclimbsearch" is taking less than 1 minute to draw bn with more accuracy , I felt it is related to my question 1.

There might be some optimization issues. If you can create an issue and share your code, I might be able to help with that.

ankurankan committed:
Fixes all issues

ankurankan committed:
Removes backport lru cache since we have dropped support for python 2

ankurankan committed:
Remove extern.lru_cache dependencies

ankurankan committed:
Missing file

ankurankan commented:
Part of this has been merged. For the rest (parallel computation) ref #940.
